### DISCUSSION TITLE
Integrated scancard generator

### DISCUSSION THREAD

**Discussion by aaaidan** [HomeSpan/HomeSpan#38] [Ideas]
I just made this homekit scan card generator. 
https://homekit-scan.surge.sh/?code=46637726

Maybe HomeSpan could adapt it to its needs. In particular, I thought the message in the CLI after you change the setup code could link to this.

```
Generating SRP verification data for new Setup Code: 466-37-726 ... New Code Saved!
Optional QR Code: X-HM://00AI1PHNMHSPN
View your scan card here: https://pages.github.com/HomeSpan/scancard?code=46637726
```

Copy paste the url, and you could get something like:
<img width="335" alt="image" src="https://user-images.githubusercontent.com/151507/107832487-620b2280-6d45-11eb-97e8-70c055a76d72.png">

We could host it at this repo using GitHub pages. Let me know if a PR would be welcome.

---

Similar to this, it might be nice to be able to link to a webpage that displays the QR code described by `Optional QR Code: X-HM:/...`. If QR is better than "Scancardium", we could do a similar thing with QR Codes instead.

**Comment by HomeSpan** [comment]
Thanks very much for sharing this.  I really like the idea of turn-key scripts to create Setup tags or QR codes, but for a variety of reasons I'd rather not add any links directly into the HomeSpan library.   Can your generator be ported to something completely standalone, such as a Javascript program that could run **locally** in the user's browser?  If so, we could then include as a module in the library.

**Reply by aaaidan** [reply]
Good info, thanks. At the moment I can't imagine a way to get this working locally, at least in a way that wouldn't be more fuss than just entering the code directly! I'll think about it...

Since local is important for hs, I wonder whether the device could emit ascii "half block" characters to print the required QR code?

Like this:

```
 ▄▄▄▄▄▄▄  ▄ ▄▄ ▄▄▄▄▄▄▄ 
 █ ▄▄▄ █ ▄█▀█▄ █ ▄▄▄ █ 
 █ ███ █  ▀▄▀▄ █ ███ █ 
 █▄▄▄▄▄█ ▄ ▄ █ █▄▄▄▄▄█ 
 ▄▄▄ ▄▄▄▄██▄█▀▄▄   ▄   
   ██▄ ▄██ █ ▄▄▀██▄▄██ 
  █▄▄ ▄▄▄█ ▄ ▄▄ ▄█  ▀  
 ▄▄▄▄▄▄▄ █▀ █ ▄ ▄█ ███ 
 █ ▄▄▄ █ ██▀█▀▀▄▀▄     
 █ ███ █ ▄▀  ▄ ▀▀█ █▀▄ 
 █▄▄▄▄▄█ █▀▀ ▄ █▄▀  █▄ 
```

This text is styled badly on github: the line height is too tall for the reader to recognise the code. 

But here's what it looks like in the terminal I use (with PlatformIO), and it actually works! 😄 
![image](https://user-images.githubusercontent.com/151507/109099179-82d55f80-76d7-11eb-911a-53d0240ab7f8.png)

Not sure if it works in Arduino's serial monitor, but I'll check soon.

The other side of the equation that I don't know much about is how much this would increase binary sizes and compile times. Linking in a library like `libqrencode` would be super fun _and_ functional, but would obviously add to the build size. Could be worth it if it's lightweight, no? What would be acceptable?

Thoughts?

**Reply by aaaidan** [reply]
Oh, cool. If you put the `X-HM://00AI1PHNMHSPN` as text in the QR code, the scanner recognizes to open it in the Home app, and shows a helpful notification.

<img src="https://user-images.githubusercontent.com/151507/109100143-33902e80-76d9-11eb-8251-2fe9c4560362.png" alt="homekit device pairing notification from qr code"  width="250" />
